Bricks/net/api/response.h

```diff
@@ -0,0 +1,155 @@
+// the best design is to return a value from one function to have the next one in the chain,
+// which can be `std::move(request)`, to pick it up.
+
+#include "request.h"
+
+#include "../http/http.h"
+#include "../../strings/is_string_type.h"
+#include "../../cerealize/cerealize.h"
+
+namespace bricks {
+namespace net {
+namespace api {
+
+static_assert(strings::is_string_type<const char*>::value, "");
+static_assert(!strings::is_string_type<int>::value, "");
+
+struct Response {
+  std::string body;
+  HTTPResponseCodeValue code;
+  std::string content_type;
+  HTTPHeadersType extra_headers;
+
+  Response() = delete;
+  void operator=(const Response&) = delete;
+  void operator=(Response&&) = delete;
+
+  template <typename... ARGS>
+  Response(ARGS&&... args) {
+    Construct(std::forward<ARGS>(args)...);
+  }
+
+  void Construct(const Response& rhs) {
+    body = rhs.body;
+    code = rhs.code;
+    content_type = rhs.content_type;
+    extra_headers = rhs.extra_headers;
+  }
+
+  void Construct(Response&& rhs) {
+    body = std::move(rhs.body);
+    code = rhs.code;
+    content_type = rhs.content_type;
+    extra_headers = rhs.extra_headers;
+  }
+
+  template <typename T>
+  typename std::enable_if<!std::is_same<bricks::decay<T>, Response>::value &&
+                          strings::is_string_type<T>::value>::type
+  Construct(T&& body,
+            HTTPResponseCodeValue code = HTTPResponseCode.OK,
+            const std::string& content_type = HTTPServerConnection::DefaultContentType(),
+            const HTTPHeadersType& extra_headers = HTTPHeadersType()) {
+    this->body = std::forward<T>(body);
+    this->code = code;
+    this->content_type = content_type;
+    this->extra_headers = extra_headers;
+  }
+
+  template <typename T>
+  typename std::enable_if<!std::is_same<bricks::decay<T>, Response>::value &&
+                          !(strings::is_string_type<T>::value)>::type
+  Construct(T&& object,
+            HTTPResponseCodeValue code = HTTPResponseCode.OK,
+            const HTTPHeadersType& extra_headers = HTTPHeadersType()) {
+    this->body = cerealize::JSON(std::forward<T>(object)) + '\n';
+    this->code = code;
+    this->content_type = HTTPServerConnection::DefaultJSONContentType();
+    this->extra_headers = extra_headers;
+  }
+
+  template <typename T>
+  typename std::enable_if<!std::is_same<bricks::decay<T>, Response>::value &&
+                          !(strings::is_string_type<T>::value)>::type
+  Construct(T&& object,
+            const std::string& object_name,
+            HTTPResponseCodeValue code = HTTPResponseCode.OK,
+            const HTTPHeadersType& extra_headers = HTTPHeadersType()) {
+    this->body = cerealize::JSON(std::forward<T>(object), object_name) + '\n';
+    this->code = code;
+    this->content_type = HTTPServerConnection::DefaultJSONContentType();
+    this->extra_headers = extra_headers;
+  }
+
+  Response& Body(const std::string& s) {
```

Comment:
With empty (default) initialization, not without initialization. Otherwise, these setters look redundant to me as we've got every field initialized via the constructing functions above.